File: src/cpp/tests/test_presets.cpp
#include <gtest/gtest.h>

#include <algorithm>
#include <cstdio>
#include <string>
#include <unistd.h>

#include "bcc950/constants.hpp"
#include "bcc950/presets.hpp"

namespace bcc950 {
namespace {

/// Fixture that provides a unique temporary file for each test.
class PresetsTest : public ::testing::Test {
protected:
    std::string tmp_path;

    void SetUp() override {
        tmp_path = "/tmp/bcc950_test_presets_"
                   + std::to_string(getpid()) + "_"
                   + std::to_string(reinterpret_cast<uintptr_t>(this))
                   + ".json";
    }

    void TearDown() override {
        std::remove(tmp_path.c_str());
    }
};

// ---- SaveAndRecall ----

TEST_F(PresetsTest, SaveAndRecallSinglePreset) {
    PresetManager pm(tmp_path);

    PositionTracker pos;
    pos.pan  = 2.5;
    pos.tilt = -1.0;
    pos.zoom = 350;

    pm.save_preset("home", pos);

    auto result = pm.recall_preset("home");
    ASSERT_TRUE(result.has_value());
    EXPECT_DOUBLE_EQ(result->pan, 2.5);
    EXPECT_DOUBLE_EQ(result->tilt, -1.0);
    EXPECT_EQ(result->zoom, 350);
}

TEST_F(PresetsTest, RecallNonexistentReturnsNullopt) {
    PresetManager pm(tmp_path);
    auto result = pm.recall_preset("does_not_exist");
    EXPECT_FALSE(result.has_value());
}

TEST_F(PresetsTest, SaveOverwritesExistingPreset) {
    PresetManager pm(tmp_path);

    PositionTracker pos1;
    pos1.pan = 1.0;
    pos1.tilt = 1.0;
    pos1.zoom = 200;
    pm.save_preset("spot", pos1);

    PositionTracker pos2;
    pos2.pan = -2.0;
    pos2.tilt = -2.0;
    pos2.zoom = 400;
    pm.save_preset("spot", pos2);

    auto result = pm.recall_preset("spot");
    ASSERT_TRUE(result.has_value());
    EXPECT_DOUBLE_EQ(result->pan, -2.0);
    EXPECT_DOUBLE_EQ(result->tilt, -2.0);
    EXPECT_EQ(result->zoom, 400);

    // Still only one preset
    EXPECT_EQ(pm.list_presets().size(), 1u);
}

// ---- DeletePreset ----

TEST_F(PresetsTest, DeletePresetRemovesIt) {
    PresetManager pm(tmp_path);

    PositionTracker pos;
    pos.pan = 1.0;
    pm.save_preset("temp", pos);

    EXPECT_TRUE(pm.delete_preset("temp"));
    EXPECT_FALSE(pm.recall_preset("temp").has_value());
    EXPECT_TRUE(pm.list_presets().empty());
}

TEST_F(PresetsTest, DeleteNonexistentReturnsFalse) {
    PresetManager pm(tmp_path);
    EXPECT_FALSE(pm.delete_preset("nonexistent"));
}

TEST_F(PresetsTest, DeleteDoesNotAffectOtherPresets) {
    PresetManager pm(tmp_path);

    PositionTracker pos;
    pos.pan = 1.0;
    pm.save_preset("keep", pos);

    pos.pan = 2.0;
    pm.save_preset("remove", pos);

    EXPECT_TRUE(pm.delete_preset("remove"));
    EXPECT_EQ(pm.list_presets().size(), 1u);

    auto kept = pm.recall_preset("keep");
    ASSERT_TRUE(kept.has_value());
    EXPECT_DOUBLE_EQ(kept->pan, 1.0);
}

// ---- ListPresets ----

TEST_F(PresetsTest, ListPresetsEmptyByDefault) {
    PresetManager pm(tmp_path);
    EXPECT_TRUE(pm.list_presets().empty());
}

TEST_F(PresetsTest, ListPresetsReturnsAllNames) {
    PresetManager pm(tmp_path);

    PositionTracker pos;
    pm.save_preset("alpha", pos);
    pm.save_preset("beta", pos);
    pm.save_preset("gamma", pos);

    auto names = pm.list_presets();
    EXPECT_EQ(names.size(), 3u);

    // Verify all names are present (order may vary)
    std::sort(names.begin(), names.end());
    EXPECT_EQ(names[0], "alpha");
    EXPECT_EQ(names[1], "beta");
    EXPECT_EQ(names[2], "gamma");
}

TEST_F(PresetsTest, ListPresetsAfterDelete) {
    PresetManager pm(tmp_path);

    PositionTracker pos;
    pm.save_preset("one", pos);
    pm.save_preset("two", pos);
    pm.save_preset("three", pos);

    pm.delete_preset("two");

    auto names = pm.list_presets();
    EXPECT_EQ(names.size(), 2u);

    std::sort(names.begin(), names.end());
    EXPECT_EQ(names[0], "one");
    EXPECT_EQ(names[1], "three");
}

// ---- PersistenceAcrossInstances ----

TEST_F(PresetsTest, PersistenceAcrossInstances) {
    // Save presets with one instance
    {
        PresetManager pm1(tmp_path);

        PositionTracker p1;
        p1.pan  = 3.0;
        p1.tilt = -1.5;
        p1.zoom = 250;
        pm1.save_preset("desk", p1);

        PositionTracker p2;
        p2.pan  = -2.0;
        p2.tilt = 1.0;
        p2.zoom = 400;
        pm1.save_preset("window", p2);
    }
    // pm1 is destroyed here; data should have been persisted to disk

    // Load presets with a new instance
    {
        PresetManager pm2(tmp_path);

        auto names = pm2.list_presets();
        EXPECT_EQ(names.size(), 2u);

        auto desk = pm2.recall_preset("desk");
        ASSERT_TRUE(desk.has_value());
        EXPECT_DOUBLE_EQ(desk->pan, 3.0);
        EXPECT_DOUBLE_EQ(desk->tilt, -1.5);
        EXPECT_EQ(desk->zoom, 250);

        auto window = pm2.recall_preset("window");
        ASSERT_TRUE(window.has_value());
        EXPECT_DOUBLE_EQ(window->pan, -2.0);
        EXPECT_DOUBLE_EQ(window->tilt, 1.0);
        EXPECT_EQ(window->zoom, 400);
    }
}

TEST_F(PresetsTest, PersistenceDeleteSurvivesReload) {
    // Save a preset, then delete it, then reload
    {
        PresetManager pm1(tmp_path);
        PositionTracker pos;
        pos.pan = 1.0;
        pm1.save_preset("ephemeral", pos);
        pm1.save_preset("permanent", pos);
        pm1.delete_preset("ephemeral");
    }
    {
        PresetManager pm2(tmp_path);
        EXPECT_FALSE(pm2.recall_preset("ephemeral").has_value());
        EXPECT_TRUE(pm2.recall_preset("permanent").has_value());
        EXPECT_EQ(pm2.list_presets().size(), 1u);
    }
}

TEST_F(PresetsTest, PersistenceMultipleRoundTrips) {
    PositionTracker pos;

    // Round 1: save
    {
        PresetManager pm(tmp_path);
        pos.pan = 1.0;
        pos.tilt = 0.5;
        pos.zoom = 200;
        pm.save_preset("spot_a", pos);
    }

    // Round 2: add another, verify first still there
    {
        PresetManager pm(tmp_path);
        ASSERT_TRUE(pm.recall_preset("spot_a").has_value());

        pos.pan = -1.0;
        pos.tilt = -0.5;
        pos.zoom = 300;
        pm.save_preset("spot_b", pos);
    }

    // Round 3: verify both
    {
        PresetManager pm(tmp_path);
        EXPECT_EQ(pm.list_presets().size(), 2u);

        auto a = pm.recall_preset("spot_a");
        ASSERT_TRUE(a.has_value());
        EXPECT_DOUBLE_EQ(a->pan, 1.0);

        auto b = pm.recall_preset("spot_b");
        ASSERT_TRUE(b.has_value());
        EXPECT_DOUBLE_EQ(b->pan, -1.0);
    }
}

} // anonymous namespace
} // namespace bcc950

File: src/python/tests/unit/test_presets.py
"""Unit tests for PresetManager."""

import json
import pytest

from bcc950.presets import PresetManager
from bcc950.position import PositionTracker


# ---------------------------------------------------------------------------
# Save / Recall
# ---------------------------------------------------------------------------

class TestSaveRecall:
    def test_save_and_recall(self, tmp_path):
        pm = PresetManager(tmp_path / "presets.json")
        pos = PositionTracker(pan=1.0, tilt=-0.5, zoom=200)
        pm.save_preset("home", pos)

        recalled = pm.recall_preset("home")
        assert recalled is not None
        assert recalled.pan == pytest.approx(1.0)
        assert recalled.tilt == pytest.approx(-0.5)
        assert recalled.zoom == 200

    def test_recall_nonexistent_returns_none(self, tmp_path):
        pm = PresetManager(tmp_path / "presets.json")
        assert pm.recall_preset("nope") is None

    def test_save_overwrites_existing(self, tmp_path):
        pm = PresetManager(tmp_path / "presets.json")
        pm.save_preset("spot", PositionTracker(pan=1.0, tilt=0.0, zoom=100))
        pm.save_preset("spot", PositionTracker(pan=2.0, tilt=1.0, zoom=300))

        recalled = pm.recall_preset("spot")
        assert recalled is not None
        assert recalled.pan == pytest.approx(2.0)
        assert recalled.tilt == pytest.approx(1.0)
        assert recalled.zoom == 300


# ---------------------------------------------------------------------------
# Delete
# ---------------------------------------------------------------------------

class TestDelete:
    def test_delete_existing(self, tmp_path):
        pm = PresetManager(tmp_path / "presets.json")
        pm.save_preset("temp", PositionTracker())
        assert pm.delete_preset("temp") is True
        assert pm.recall_preset("temp") is None

    def test_delete_nonexistent(self, tmp_path):
        pm = PresetManager(tmp_path / "presets.json")
        assert pm.delete_preset("nope") is False


# ---------------------------------------------------------------------------
# List
# ---------------------------------------------------------------------------

class TestList:
    def test_list_empty(self, tmp_path):
        pm = PresetManager(tmp_path / "presets.json")
        assert pm.list_presets() == []

    def test_list_after_saves(self, tmp_path):
        pm = PresetManager(tmp_path / "presets.json")
        pm.save_preset("a", PositionTracker())
        pm.save_preset("b", PositionTracker())
        pm.save_preset("c", PositionTracker())
        assert sorted(pm.list_presets()) == ["a", "b", "c"]

    def test_list_after_delete(self, tmp_path):
        pm = PresetManager(tmp_path / "presets.json")
        pm.save_preset("keep", PositionTracker())
        pm.save_preset("remove", PositionTracker())
        pm.delete_preset("remove")
        assert pm.list_presets() == ["keep"]


# ---------------------------------------------------------------------------
# Persistence
# ---------------------------------------------------------------------------

class TestPersistence:
    def test_survives_new_instance(self, tmp_path):
        """Presets saved by one manager instance should be loadable by another."""
        path = tmp_path / "presets.json"

        pm1 = PresetManager(path)
        pm1.save_preset("office", PositionTracker(pan=2.0, tilt=1.0, zoom=250))

        # Create a completely new manager instance pointing at the same file
        pm2 = PresetManager(path)
        recalled = pm2.recall_preset("office")
        assert recalled is not None
        assert recalled.pan == pytest.approx(2.0)
        assert recalled.tilt == pytest.approx(1.0)
        assert recalled.zoom == 250

    def test_file_is_valid_json(self, tmp_path):
        """The presets file should be valid JSON."""
        path = tmp_path / "presets.json"
        pm = PresetManager(path)
        pm.save_preset("test", PositionTracker(pan=0.5, tilt=-0.5, zoom=150))

        with open(path) as f:
            data = json.load(f)

        assert "test" in data
        assert data["test"]["pan"] == pytest.approx(0.5)
        assert data["test"]["tilt"] == pytest.approx(-0.5)
        assert data["test"]["zoom"] == 150

    def test_delete_persists(self, tmp_path):
        """Deleting a preset should persist the deletion to disk."""
        path = tmp_path / "presets.json"

        pm1 = PresetManager(path)
        pm1.save_preset("gone", PositionTracker())
        pm1.delete_preset("gone")

        pm2 = PresetManager(path)
        assert pm2.recall_preset("gone") is None
        assert "gone" not in pm2.list_presets()

    def test_multiple_presets_persist(self, tmp_path):
        path = tmp_path / "presets.json"

        pm1 = PresetManager(path)
        pm1.save_preset("p1", PositionTracker(pan=1.0, tilt=0.0, zoom=100))
        pm1.save_preset("p2", PositionTracker(pan=-1.0, tilt=2.0, zoom=400))

        pm2 = PresetManager(path)
        assert sorted(pm2.list_presets()) == ["p1", "p2"]
        r1 = pm2.recall_preset("p1")
        r2 = pm2.recall_preset("p2")
        assert r1.pan == pytest.approx(1.0)
        assert r2.pan == pytest.approx(-1.0)
        assert r2.zoom == 400
